drivers/i3c/i3c_i2c_virtual_controller.c

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2020 Google LLC
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ * Copyright 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT virtual_i3c_i2c_controller
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(i3c_i2c_virtual_controller, CONFIG_I3C_LOG_LEVEL);
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/i3c.h>
+
+struct i2c_virt_ctrl_data {
+	/* Pointer to the array of I2C devices */
+	struct i3c_i2c_device_desc *i2c_devs;
+
+	/* Number of I2C devices */
+	const size_t num_i2c_devs;
+};
+
+static struct i3c_i2c_device_desc *find_i2c_dev(const struct i2c_virt_ctrl_data *data,
+						uint8_t addr)
+{
+	int i;
+
+	for (i = 0; i < data->num_i2c_devs; i++) {
+		if (data->i2c_devs[i].addr == addr) {
+			return &data->i2c_devs[i];
+		}
+	}
+
+	return NULL;
+}
+
+static int i2c_virt_ctrl_configure(const struct device *dev, uint32_t dev_config)
+{
+	return 0;
+}
+
+static int i2c_virt_ctrl_transfer(const struct device *dev, struct i2c_msg *msgs,
+				  uint8_t num_msgs, uint16_t addr)
+{
+	const struct i2c_virt_ctrl_data *data = dev->data;
+	struct i3c_i2c_device_desc *i2c_dev;
+
+	i2c_dev = find_i2c_dev(data, addr);
+
+	if (i2c_dev == NULL) {
+		LOG_ERR("Cannot find I2C device 0x%02x", addr);
+		return -ENODEV;
+	}
+
+	return i3c_i2c_transfer(i2c_dev, msgs, num_msgs);
+}
+
+/**
+ * Set up a new emulator and add it to the list
+ *
+ * @param dev I2C emulation controller device
+ */
+static int i2c_virt_ctrl_init(const struct device *dev)
+{
+	const struct i2c_virt_ctrl_data *data = dev->data;
+
+	int i;
+
+	for (i = 0; i < data->num_i2c_devs; i++) {
+		i3c_i2c_device_register(&data->i2c_devs[i]);
+	}
+
+	return 0;
+}
+
+static struct i2c_driver_api i2c_virt_ctrl_api = {
+	.configure = i2c_virt_ctrl_configure,
+	.transfer = i2c_virt_ctrl_transfer,
+};
+
+#define I2C_DEVICE_DESC(node_id)					\
+	{								\
+		.bus = DEVICE_DT_GET(DT_BUS(DT_PARENT(node_id))),	\
+		.addr = DT_PROP_BY_IDX(node_id, reg, 0),		\
+		.lvr = DT_PROP(DT_PARENT(node_id), lvr),		\
+	}
+
+#define I2C_VIRT_CTRL_INIT(n)						\
+	static struct i3c_i2c_device_desc i2c_dev_list_##n[] = {	\
```

Comment:
shouldn't this be `const`. I wouldn't expect the number of i2c devices to change a run time and it should all be known at compile time.